pkg/kubelet/volume/iscsi_pd/iscsi_util.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iscsi_pd
+
+import (
+	"errors"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/mount"
+	"github.com/golang/glog"
+)
+
+func probeDevicePath(devicePath string, maxRetries int) bool {
+	numTries := 0
+	for {
+		_, err := os.Stat(devicePath)
+		if err == nil {
+			return true
+		}
+		if err != nil && !os.IsNotExist(err) {
+			return false
+		}
+		numTries++
+		if numTries == maxRetries {
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	return false
+}
+
+func makePDNameInternal(host volume.Host, portal string, iqn string, lun string) string {
+	return path.Join(host.GetPluginDir(ISCSIDiskPluginName), "iscsi", portal, iqn, "lun", lun)
+}
+
+type ISCSIDiskUtil struct{}
+
+func (util *ISCSIDiskUtil) MakeGlobalPDName(disk interface{}) string {
+	iscsi := disk.(iscsiDisk)
+	return makePDNameInternal(iscsi.plugin.host, iscsi.portal, iscsi.iqn, iscsi.lun)
+}
+
+func (util *ISCSIDiskUtil) AttachDisk(disk interface{}) error {
+	iscsi := disk.(iscsiDisk)
+	devicePath := strings.Join([]string{"/dev/disk/by-path/ip", iscsi.portal, "iscsi", iscsi.iqn, "lun", iscsi.lun}, "-")
+	exist := probeDevicePath(devicePath, 1)
+	if exist == false {
+		// discover iscsi target
+		_, err := iscsi.execCommand("iscsiadm", []string{"-m", "discovery", "-t", "sendtargets", "-p",
```

Comment:
I'm 99% sure it will return an error - if iscsiadm has a -v flag or something that is ideal.